muon/_core/preproc.py

```diff
@@ -155,9 +155,25 @@ def _l2norm(
     adata: AnnData, rep: Optional[Union[Iterable[str], str]] = None, n_pcs: Optional[int] = 0
 ):
     X = _choose_representation(adata, rep, n_pcs)
-    norm = X / np.linalg.norm(X, ord=2, axis=1, keepdims=True)
-    norm[~np.isfinite(norm)] = 0
+    sparse_X = issparse(X)
+    if sparse_X:
+        X_norm = linalg.norm(X, ord=2, axis=1)
+        norm = X / np.expand_dims(X_norm, axis=1)
+        if issparse(norm):
+            # find nan and infinite values and construct matrix without
+            i, j, val = find(norm)
+            isfin = np.isfinite(val)
+            i = i[isfin]
+            j = j[isfin]
+            val = val[isfin]
+            norm = csr_matrix((val, (i, j)), shape=X.shape)
+    else:
+        norm = X / np.linalg.norm(X, ord=2, axis=1, keepdims=True)
+        norm[~np.isfinite(norm)] = 0
     X.astype(norm.dtype, copy=False)
+    if sparse_X and not issparse(norm):
+        norm[~np.isfinite(norm)] = 0
+        X = X.toarray()
```

Comment:
I don't think that will do what you're expecting it to do. The line below (`X[:] = norm`) overwrites the matrix in the AnnData object with the normalized values. Now you're replacing the local X variable with something else, so the normalized values will be lost. The question is: Is there ever a situation where `X` is sparse and `norm` is not? At least if `X` is a `csr_matrix` or a `csc_matrix` `norm` [should always be a COO matrix](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.coo_array.html#scipy.sparse.coo_array).

Then you could also get rid of the if clause in line 162.

Reply:
Thanks for the feedback. I removed the lines in commit https://github.com/scverse/muon/pull/114/commits/e7b7b12ace6f77086c06aa1194a8159060004645